### PR TITLE
Find the correct fields when using page filter on `struct` fields in parquet

### LINKED ISSUE #8456: query result empty  when a struct field name and a regular field name is same
### Describe the bug

When a struct field name is the same as a regular field name, and the struct is declared before the regular field, it can cause issues when the field name in where clause

### To Reproduce

construct a parquet file ,write 2 records  as follow
```
    +---------------------+----+--------+
     | struct              | id | name   |
     +---------------------+----+--------+
     | {id: 1, name: aaa1} | 1  | test01 |
     | {id: 2, name: aaa2} | 2  | test02 |
     +---------------------+----+--------+
```
 when execute sql `select * from base_table where name='test01'` ,will got empty result
here is the bug demo
https://github.com/anlihust/datafusion_demo/blob/main/src/main.rs

### Expected behavior

 when execute sql `select * from base_table where name='test01'` ,get the test01 record
```
    +---------------------+----+--------+
     | struct              | id | name   |
     +---------------------+----+--------+
     | {id: 1, name: aaa1} | 1  | test01 |
     +---------------------+----+--------+
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by manoj-inukolunu**
## Which issue does this PR close?

Closes #8456

## What changes are included in this PR?

The issue is in this line  `c.column_descr().name() == column.name()` in `find_column_index` `datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs` . When struct fields have the same name as the column names that condition is returning the struct field instead of the column field.

## Are these changes tested?
Yes I have added a test which is taken from the reproduction code.

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs [hunk 1]
```diff
@@ -280,7 +280,7 @@ fn find_column_index(
         .columns()
         .iter()
         .enumerate()
-        .find(|(_idx, c)| c.column_descr().name() == column.name())
+        .find(|(_idx, c)| c.column_descr().name() == column.name() && c.column_descr().path().parts().len() == 0)
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs:283]: Perhaps you can use the `parquet_column` function

https://github.com/apache/arrow-datafusion/blob/06bbe1298fa8aa042b6a6462e55b2890969d884a/datafusion/core/src/datasource/physical_plan/parquet/statistics.rs#L133

Which is how the row group metadata fetches the relevant metadata
https://github.com/apache/arrow-datafusion/blob/eb81ea299aa7e121bbe244e7e1ab56513d4ef800/datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs#L252-L255

**Review by alamb — COMMENTED** [review]
Hi @manoj-inukolunu thank you for this contribution

There appears to be a CI failure as well
https://github.com/apache/arrow-datafusion/actions/runs/7510755702/job/20489197785?pr=8848

File: datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs (hunk 1 above)
Comment (manoj-inukolunu) [reply datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs]: Update this PR @alamb .

File: datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs [hunk 2]
```diff
@@ -164,17 +169,23 @@ impl PagePruningPredicate {
...
         for predicate in page_index_predicates {
             // find column index by looking in the row group metadata.
-            let col_idx = find_column_index(predicate, &groups[0]);
 
+            let name = find_column_name(predicate);
+            let mut parquet_col = None;
+            if name.is_some() {
+                parquet_col =
+                    parquet_column(parquet_schema, arrow_schema, name.unwrap().as_str());
+            }
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs:178]: I think a more rust ideomatic way of doing this pattern that avoids the unwrap and mut is something like

```suggestion
            let parquet_col = name
              .and_then(|name| {
                    parquet_column(parquet_schema, arrow_schema, name.as_str());
            });
```

File: datafusion/core/src/datasource/physical_plan/parquet/mod.rs [hunk 3]
```diff
@@ -2136,4 +2144,65 @@ mod tests {
         let execution_props = ExecutionProps::new();
         create_physical_expr(expr, &df_schema, &execution_props).unwrap()
     }
+
+    #[tokio::test]
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/mod.rs:2148]: I verified that this test fails without the code in this PR 👍 

```
thread 'datasource::physical_plan::parquet::tests::test_struct_filter_parquet' panicked at datafusion/core/src/datasource/physical_plan/parquet/mod.rs:2156:9:
assertion `left == right` failed
  left: 0
 right: 1
stack backtrace:
```

File: datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs [hunk 4]
```diff
@@ -249,10 +260,7 @@ impl PagePruningPredicate {
 /// that `extract_page_index_push_down_predicates` only return
 /// predicate with one col)
 ///
-fn find_column_index(
-    predicate: &PruningPredicate,
-    row_group_metadata: &RowGroupMetaData,
-) -> Option<usize> {
+fn find_column_name(predicate: &PruningPredicate) -> Option<String> {
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs:263]: This function docstring needs to be updated if we make this change. However I have a slightly different proposal here: https://github.com/manoj-inukolunu/arrow-datafusion/pull/1

**Review by alamb — COMMENTED** [review]
Thank you so much @manoj-inukolunu  -- this looks great

I have some style suggestions in this proposal which targets your branch: https://github.com/manoj-inukolunu/arrow-datafusion/pull/1

However they aren't required -- if you would prefer I think we could just update the docstrings for `find_column_name` instead

Thanks again 🙏

File: datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs (hunk 2 above)
Comment (manoj-inukolunu) [reply datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs]: aah thank you , i was searching for something like this. I will add this to my toolbox.

File: datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs (hunk 4 above)
Comment (manoj-inukolunu) [reply datafusion/core/src/datasource/physical_plan/parquet/page_filter.rs]: Thank you @alamb  I like the proposal its much cleaner. I have merged your pr.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
I took the liberty of merging this branch with main and running `cargo fmt` which will hopefully get the CI to pass

**Comment by alamb** [comment]
Thanks again @manoj-inukolunu

**Pull request merged**